lightning_pose/data/datasets.py

```diff
@@ -335,3 +339,184 @@ def __getitem__(self, idx: int) -> HeatmapLabeledExampleDict:
+        self.data_length = self.data_length.pop()
+
+        # self.num_keypoints_unique = self.num_keypoints.pop()
+        # self.num_keypoints = self.num_keypoints_unique * self.num_views
+        self.num_targets = self.num_keypoints * 2
+
+        # check if all the data is in correct order
+        self.check_data_images_names()
+
+    def check_data_images_names(self, delimiter: str = "img"):
+        """Data checking
+        Each object in self.datasets will have the attribute image_names
+        (i.e. self.datasets['top'].image_names) since each values is a
+        HeatmapDataset. Include a check to make sure that the image names
+        are the same across all views, so that when it loads element n from
+        each individual view we know these are properly matched.
+
+        Args:
+
+            delimiter: for spliting the file name string to get the frame number and format.
+
+        """
+
+        for idx in range(self.data_length):
+            img_name_buff = []
+            for view, heatmaps in self.dataset.items():
+                img_name_buff.append(heatmaps.image_names[idx].split(delimiter)[-1])
+                if len(set(img_name_buff)) != 1:
+                    raise ImportError(f"Discrepancy in images names across CSV \
+                                      files! index:{idx}, image frame names:{img_name_buff}")
+
+    @property
+    def height(self) -> int:
+        # is this correct?
+        # shouldn't it be self.dataset[self.view_names[0]].height * self.num_views ?
+        return self.dataset[self.view_names[0]].height * self.num_views
+
+    @property
+    def width(self) -> int:
+        # assume resizing transformation is the last imgaug one
+        return self.imgaug_transform[-1].get_parameters()[0][1].value
+
+    def __len__(self) -> int:
+        return self.data_length
+
+    @property
+    def output_shape(self) -> tuple:
+        return (
+            self.height // 2**self.downsample_factor,
+            self.width // 2**self.downsample_factor,
+        )
+
+    @property
+    def num_views(self) -> int:
+        return len(self.view_names)
+
+    def fusion(self, datadict: dict, bbx: Optional[np.array] = np.array((0))):
+        """ Here all the view data will be merged into one image and multiple heatmaps.
+            images and heatmaps will be concatenated vertically dim=1
+        args:
+            heatmaps: this comes from HeatmapDataset.__getItems__(idx) for each view.
+            bbx: the part of the image that needs to be cropped
+        """
+        heatmaps = []
+        keypoints = []
+        images = []
+        concat_order = []
+        for view_num, (view, data) in enumerate(datadict.items()):
+            heatmap_key_num = data["heatmaps"].shape[0]
+            heatmap_length = data["heatmaps"].shape[1]
+            heatmap_width = data["heatmaps"].shape[2]
+            heatmap_buffer = torch.zeros(heatmap_key_num,
```

Comment:
this part can be simplified - basically, we just want to concatenate along the keypoint dimension. so if view 0 has 4 keypoints and view 1 has 8 keypoints, then the final heatmap tensor just needs to be of shape (12, heatmap_height, heatmap_width) where 12 = 4 + 8; same with keypoints.

so you should be able to do something like
```
heatmaps = torch.cat([d["heatmaps"] for _, d in datadict.items()])
```

for the images, even if they start with a different initial size we are going to resize them all the same. so the final image tensor will be of shape (num_views, 3, image_height, image_width)